### DISCUSSION TITLE
What are the security use cases you need covered by OpenSearch Dashboards Security out of the box?

### DISCUSSION THREAD

**Discussion by shanilpa** [opensearch-project/ux#54] [General]
## **Context**
Currently, OpenSearch Dashboards does not have built in Security capabilities (Security is managed through the security plugin). This leads to several issues and limitations for creators and users. In the future OpenSearch Dashboards Core will develop security as a first class citizen. In order to build a holistic solution we are looking to gather a range of use cases from users, developers, and community members to help inform the development of this feature. 

If the current security plugin provides the feature already that's okay mention it again so we can keep track of it.
If you would like additional or new security capabilities for OpenSearch Dashboards mention those as well.

**NB: Do not use this discussion to post issues, request support, or ask specific feature questions. [The OpenSearch Forum](https://forum.opensearch.org/) or appropriate Github repo might be right place for those.**

### Use Case Title: 
Simple yet descriptive. 

### Context
- Any additional information you can provide that helps us understand the use cases you mentioned.

### Use Case
_If possible, please provide your suggestions as a user story. Be as specific as possible:_
-  **Example Template:** As a [[ user ]] I want to be able to [[ action ]] so that I can [[ outcome ]].
- **Example**: As an [[ administrator that manages my teams OpenSearch Dashboards ]] I want to be able to [[ quickly create users and grant them permissions ]] so that I can [[ give them access to the things they need to do their jobs. ]]

**Comment by shanilpa** [comment]
## Allowing individual users to personalize some aspects of OS Dashboards

### **Context**
Currently, OpenSearch Dashboards doesn't have a good way to understand who the user is. Therefore tasks like personalization are difficult to implement and users inherit settings made by an administrator for the entirety of their Dashboards instance. Though this global application of settings is useful in some use cases for others it is a hindrance.

In the future it would be awesome to be able to offer personalization by mapping specific configuration features to users.

### **User Cases**
1. As [[ any Dashboards user ]] I want to be able to [[ control the accessibility and theming of my instance of dashboards ]] so that I can [[ setup in a way that works best for me and my productivity ]].

**Reply by shanilpa** [reply]
@KrooshalUX I have theming captured here. Feel free to add any specific items in the comments

**Comment by shanilpa** [comment]
## Configuring Multi-tenancy through UI

### Context
Multi-tenancy in Dashboards is a central concept if the security plugin is enabled. However a good amount of configuration related to it is handled through YAML files. To offer a better user experience these configurations should be made available through a UI.

### Use Cases

1. As a [[ OS Dashboard's admin user ]] I want to be able to [[ disable the global and or private tenant through a UI]] so that I can [[ have control over the usage of tenants in my instance of Dashboards without editing config files ]].
2. As a [[ OSDashboard's admin user ]] I want to be able to [[ set a default tenant through the UI ]] so that I can [[ have control over the usage of tenants in my instance of Dashboards ]].

**Reply by seraphjiang** [reply]
> Multi-tenancy in Dashboards is a central concept if the security plugin is enabled. However a good amount of configuration related to it is handled through YAML files. To offer a better user experience these configurations should be made available through a UI.

tenancy is introduced only with security plugin. it is not typical feature or concept, user could easily understand. This legacy should not be included in security out of box project scope. To support user still need it, they could still use existing security plugin.

**Reply by shanilpa** [reply]
From a UX perspective having users manage security both in the Security Plugin and via security out of the box would lead to a frustrating and potentially confusing experience. If we are to phase out the security plugin and fully integrate it into Dashboards core we should look to find a home for Multi-tenancy as a concept somewhere in Dashboards (or as a separate plugin / extension).

My proposition here is to rethink the concept of multi-tenancy and make it a Dashboards core feature, however the security of this feature will be handled by Security out of the box. I feel this would offer the best UX without having Multi-tenancy be a feature part of the Security out of the Box scope.

**Comment by shanilpa** [comment]
## Actioning and requesting permissions in Dashboards

### Context
Currently OpenSearch Dashboards does not support permission requests by users. These requests have to be made offline to their admin who then has to do some digging and leg work to be able to grant the right level of access.

### Use Cases

1. As a [[ Dashboards user ]] I want to be able to [[ request permission for certain features or Dashboard objects through Dashboards]] so that I can [[ efficiently do my job and not have to rely on back channeling to request permissions ]].
2. As a [[ Dashboards user / team leader ]] I want to be able to [[ quickly triage, respond, and action on permission requests in OS Dashboards for Dashboard Objects I own ]] so that I can [[ support team members in a timely manner and without having to backchannel to an admin ]].
3. As a [[ OS Dashboard's admin user ]] I want to be able to [[ quickly triage, respond, and action on permission requests in OS Dashboards]] so that I can [[ have full context on the request and grant the appropriate permissions in an efficient manner ]].

**Reply by shanilpa** [reply]
This is also reliant on a notification system / management solution which would require consideration from a security perspective. https://github.com/opensearch-project/ux/discussions/54#discussioncomment-4461454 this discussion is related to notifications.

**Comment by shanilpa** [comment]
## Provide users with notifications relevant to them

### Context
The concept of a user is not well defined in Dashboards. As a result users have a hard time accessing contextual information only important to them or their team. For example toast notifications are presented to all Dashboard users, but this notification may only be relevant to the user that triggered it. 

### Use cases
1.  As a [[ Dashboards user ]] I want to [[ be able to receive contextual notifications relevant to only me ]] so that I [[ do not distract other users with things that are not relevant to them]]
2. As a [[ Dashboards user ]] I want to [[ receive notifications relevant to all members ]] so that I can [[ be aware of team wide information ]]
3. As a [[ Dashboards user ]] I want a [[ central location to manage my notifications ]] so that I can [[ manage, triage and action on them without being overwhelmed]]
4. As a [[ Dashboards user ]] I want to [[ get contextual information ]] on async task statuses or progress [[ so that I can manage and keep track of things I have initiated in Dashboards]]

**Reply by shanilpa** [reply]
How do we handle team notifications if a user does not have access to a particular feature, or dashboard object?

**Reply by seraphjiang** [reply]
we might treat notification as any other feature, share same role access control model

**Comment by shanilpa** [comment]
## Collaboration in Dashboards

### Context
Currently collaboration in Dashboards is restricted by the limited and confusing mental model created by the Multi-tenancy feature within the security plugin. Multi-tenancy is a way of separating OSDashboard objects into specific spaces, these spaces can be assigned to individual users. This concept is useful however sharing objects between tenants and managing permissions across tenants is a complicated experience that leaves a lot to be desired. 

### Use Cases
1. As a [[ Dashboard admin / creator ]] I want [[ to be able too restrict users access to Dashboard objects based on the tenant they are in ]] so that we can [[ meet compliance needs, create focused spaces, and facilitate better organization in Dashboards]]
2. As a [[ Dashboards user ]] I want to [[ be able to share Dashboards in my tenant with other users in and out of my tenant with specific viewing, and interaction permissions ]] so that we can collaborate effectively.
3. As a [[ Non-creator user ]] I want to [[ be able to access a Dashboards share link and view the visualizations created for me without having to worry about modifying the information ]] so that I can [[ safely view and consume data in a simple interface ]] 
4. As a [[ Dashboards admin user ]] I want to [[ be able to log actions and sharing in a tenant ]] so that [[ I can review and fetch data for security and compliance reasons if needed ]]

### Related Issues
Sharing in Dashboards RFC: https://github.com/opensearch-project/security/issues/1869

**Comment by shanilpa** [comment]
## Role Based Access Control for Dashboards Features

### Context
Currently, OpenSearch Dashboards does not offer Role Based Access Control for Dashboard features. Access control for index/doc's is available through the security plugin however, this does not translate well to UX elements in Dashboards. A result users can't be granted access to specific Dashboard features and an admin can not configure the experience for users. Overall, this provides a frustrating and bloated experience for users.

### Use Cases
1. As a [[ Dashboard admin / creator ]] I want [[ to be able too grant my users specific access to Dashboard features ]] so that [[ I can administer, manage, and help provide a more focused experience for Dashboard users ]]
2. As a [[Dashboard user ]] I want to [[ be able to control what features are available to certain users when I share Dashboard objects ]] so that I [[ can control their experience and allow them to focus on what I want them to see ]]
3. As a [[ Dashboard user ]] I want to [[ be able to set view, edit, share, and admin permissions to Dashboard Objects and features ]] so that I [[ can effectively and easily collaborate in Dashboards]]

### Related issues and discussions
Security Out of the Box Feature proposal: https://github.com/opensearch-project/OpenSearch-Dashboards/issues/1215
Collaboration in Dashboards discussion: https://github.com/opensearch-project/ux/discussions/54#discussioncomment-4462295

**Comment by shanilpa** [comment]
## Authentication (AuthN) and Authorization (AuthZ) Out of the Box for OS Dashboards

### Context
Today OpenSearch Dashboards does not have built in security functionality. Security is provided via the Security plugin which is comprised of both a backend and frontend plugin - both of which need to be configured to enable security in Dashboards. The process of configuring these plugins is incredibly complicated and cumbersome for Admin users. Additionally, these plugins don't support the nuanced use cases that support a great user experience in OpenSearch Dashboards due to the tight coupling to backend concepts. In order to provide this experience we should provide AuthN and AuthZ as a part of OpenSearch Dashboards.

### Use Cases
1. As a [[ Dashboard Admin user ]] I want to [[ easily setup security (AuthN and Z) for OpenSearch Dashboards in one single place ]] so that I [[ don't have to context switch and go through a complicated security setup process ]].
2. As a [[ Dashboard Admin user ]] I want to [[ be able to easily, securely, and compliantly connect Dashboards Security with Backend security ]] so that I can [[ access and grant my users access to appropriate data and backend resources]]
3. As a [[ Dashboard user ]] I want [[ a way to authenticate into OpenSearch Dashboards ]] so that I can [[ efficiently, securely, and easily access Dashboards and data I need to do my job ]]
4. As a [[ Dashboards user ]] I want [[ a way to see the permissions and access I have ]] so that I can [[ easily and intuitively understand what I can do in Dashboards ]]

### Related Issues and Discussions
- Dashboards Authentication Out of the Box Feature Proposal: https://github.com/opensearch-project/dashboards-anywhere/issues/135
- Collaboration in Dashboards discussion: https://github.com/opensearch-project/ux/discussions/54#discussioncomment-4462295
- Role Based Access Control in Dashboards discussion: https://github.com/opensearch-project/ux/discussions/54#discussioncomment-4477373
